keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md

```diff
@@ -0,0 +1,623 @@
+---
+
+# Admission Webhooks to GA
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Object selector](#object-selector)
+  * [Scope](#scope)
+  * [http for localhost](#http-for-localhost)
+  * [timeout configuration](#timeout-configuration)
+  * [Port configuration](#port-configuration)
+  * [Plumbing existing objects to delete admission](#plumbing-existing-objects-to-delete-admission)
+  * [Mutating Plugin ordering](#mutating-plugin-ordering)
+  * [Passing {Operation}Option to Webhook](#passing-operationoption-to-webhook)
+  * [AdmissionReview v1](#admissionreview-v1)
+* [V1 API](#v1-api)
+* [V1beta1 changes](#v1beta1-changes)
+* [Validations](#validations)
+* [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Post-GA tasks](#post-ga-tasks)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+Admission webhook is a way to extend kubernetes by putting hook on object creation/modification/deletion. Admission webhooks can mutate or validate the object. This feature has been Beta since Kubernetes 1.9. This document outline required steps to graduate it to GA.
+
+## Motivation
+
+Admission webhooks are currently widely used for extending kubernetes but has been in beta for 3 releases. Current set of feature requests and bug reports on the feature shows it is close to be stable and motivation of this KEP is to address those feature requests and bug reports to move this feature toward general availability (GA).
+
+### Goals
+
+Based on the user feedback, these are proposed changes to current feature to graduate it to GA:
+
+* Object Selector: Add a general selector for objects that a webhook can target.
+* Scope: Define the scope of the targeted objects, e.g. cluster vs. namespaced.
+* http for local host: relax https constrain for localhost.
+* timeout configuration: Add a configuration field to set the timeout of the webhook call.
+* port configuration: Add a configuration field to change the service port from the default 443.
+* AdmissionReview version selection: Add a API field to set the AdmissionReview versions a webhook can understand.
+* Pass existing objects to delete and delete collection admission webhooks
+* re-run mutating plugins if any webhook changed object
+* pass OperationOption (such as CreateOption/DeleteOption) to the webhook
+
+### Non-Goals
+
+Any new feature request that is not part of this KEP would not be required for GA. These are the issues considered and rejected for GA:
+
+* Can CA trust bundle being referenced via secret/configmap instead of inline? ([discussion issue](https://github.com/kubernetes/kubernetes/issues/72944))
+* Should admission webhooks have a way to authenticate API Servers? ([Discussion issue](https://github.com/kubernetes/enhancements/pull/658))
```

Comment:
@liggitt had a look that seemed like he wasn't sure. I'm ok without it for GA, but he may have a different opinion.

Reply:
Would like to hear @mikedanese's thoughts. Changes in this area are likely to take a while to percolate. If the current method (or alternatives like targeting a local/sidecar proxy that added credentials) could hold for now, I'd be happy deferring that post-GA.